src/wayland/input_method/mod.rs

```diff
@@ -0,0 +1,376 @@
+        xkb::KEYMAP_COMPILE_NO_FLAGS,
+    )
+    .unwrap();
+    let state = xkb::State::new(&keymap);
+    (keymap.get_as_string(xkb::KEYMAP_FORMAT_TEXT_V1), state)
+}
+
+/// Handle to an input method
+#[derive(Default, Debug, Clone)]
+pub struct InputMethodHandle {
+    inner: Rc<RefCell<InputMethod>>,
+}
+
+impl InputMethodHandle {
+    fn new_handle(&self, rate: i32, delay: i32, xkb_config: XkbConfig<'_>) {
+        let mut inner = self.inner.borrow_mut();
+        inner.config(rate, delay, xkb_config)
+    }
+
+    fn add_instance(&self, instance: Main<ZwpInputMethodV2>) {
+        let mut inner = self.inner.borrow_mut();
+        inner.instance = Some(instance);
+    }
+
+    fn instance_unavailable(&self) -> bool {
+        let inner = self.inner.borrow();
+        inner.instance.is_some()
+    }
+
+    fn add_keyboard(&self, keyboard: Main<ZwpInputMethodKeyboardGrabV2>) {
+        let mut inner = self.inner.borrow_mut();
+        inner.keyboard = Some(keyboard);
+        inner.send_keyboard_info();
+    }
+
+    fn add_popup_surface(&self, popup_surface: Main<ZwpInputPopupSurfaceV2>, surface: WlSurface) {
+        let mut inner = self.inner.borrow_mut();
+        inner.popup_surface_handle = Some(popup_surface);
+        inner.popup_surface = Some(surface);
+    }
+
+    /// Takes keyboard input and sends it to the input method
+    pub fn input(&self, keycode: u32, state: KeyState, serial: Serial, time: u32) {
+        let mut inner = self.inner.borrow_mut();
+        let update_modifiers = inner.key_input(keycode, state);
+        inner.input(keycode, state, serial, time, update_modifiers)
+    }
+
+    /// Indicates that an input method has grabbed a keyboard
+    pub fn keyboard_grabbed(&self) -> bool {
+        let inner = self.inner.borrow_mut();
+        inner.keyboard.is_some()
+    }
+
+    /// used to access the Main handle from text input
+    pub fn handle(&self) -> Option<Main<ZwpInputMethodV2>> {
+        self.inner.borrow().instance.clone()
+    }
+
+    /// used to access the Main popup surface handler from text input
+    pub fn popup_surface_handle(&self) -> Option<Main<ZwpInputPopupSurfaceV2>> {
+        self.inner.borrow().popup_surface_handle.clone()
+    }
+
+    /// used to access the surface for popup
+    pub fn popup_surface(&self) -> Option<WlSurface> {
+        self.inner.borrow().popup_surface.clone()
+    }
+}
+
+/// Extend [Seat] with input method specific functionality
+pub trait InputMethodSeatTrait {
+    /// Get input method seat associated with this seat
+    /// this is also used to set xkb config parameters that will be sent to the input method
+    /// Input methods need different keyboard languages for different input methods
+    /// E.g Norwegian pinyin user will want to use a nordic keyboard layout
+    /// but a Taiwanese person will write their input method with a us layout
+    fn add_input_method(
+        &self,
+        repeat_rate: i32,
+        repeat_delay: i32,
```

Comment:
These two parameters must be swapped. Right now they're the exact opposite of `add_keyboard` and that will definitely screw people up.

Reply:
Alright, done.